saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java

```diff
@@ -0,0 +1,583 @@
+ * for both sparse and dense feature vectors. Use the appropriate functions based on the representation of your feature vector.
+ * </p>
+ *
+ * @author      Abiyaz Chowdhury <chowdh2@cooper.edu>
+ */
+
+final public class MultilayerPerceptron extends Learner{
+    /**
+     * Activation function codes. When calling the {@link MultilayerPerceptron} constructor to specify activation functions, use these values to specify
+     * the activation functions.
+     */
+    private static final int SIGMOID = 1;
+    private static final int TANH = 2;
+    private static final int RELU = 3;
+    private static final int LEAKY_RELU = 4;
+    private static final int SOFTPLUS = 5;
+
+    /**
+     * Default neural network learning parameters.
+     */
+    private static double default_learning_rate = 0.1d;
+    private static double default_momentum = 0.1d;
+    private static double default_lambda = 0.00001d;
+
+    /**
+     * Default parameters for network initialization. For classification tasks, each node of the output layer corresponds
+     * to a class. The default setting below is for binary classification.
+     */
+    private static int default_num_inputs = 128;
+    private static int[] default_hidden_layers =  new int[]{};
+    private static int[] default_activation_function = {SIGMOID};
+    private static int default_num_outputs = 2;
+
+    private static int iterations = 0;
+    private static int iterations2 = 0;
+
+    public Layer[] layers;
+
+    /** Each layer contains a set of weights, gradients, and activation values. The aggregate of all layers comprise
+     * the entire neural network.
+     */
+    public class Layer {
+        public double[][] weights;
+        private double[][] gradients;
+        private double[][] prev_gradients;
+        private double[] pre_activation;
+        private double[] post_activation;
+        private double[] errors;
+        private int activation_function;
+
+        // Initializes a layer of the neural network. Weights are all initialized to 0.
+        private Layer(int num_inputs, int num_outputs, int activation_function){
+            weights = new double[num_inputs+1][num_outputs];
+            for (double[] row : weights)
+                Arrays.fill(row,0.0d);
+            gradients = new double[num_inputs+1][num_outputs];
+            for (double[] row : gradients)
+                Arrays.fill(row,0.0d);
+            prev_gradients = new double[num_inputs+1][num_outputs];
+            for (double[] row : prev_gradients)
+                Arrays.fill(row,0.0d);
+            pre_activation = new double[num_outputs];
+            Arrays.fill(pre_activation,0.0d);
+            post_activation = new double[num_outputs];
+            Arrays.fill(post_activation,0.0d);
+            errors = new double[num_outputs];
+            Arrays.fill(errors,0.0d);
+            this.activation_function = activation_function;
+        }
+
+        public void write(ExceptionlessOutputStream out){
```

Comment:
Quick Question: while serializing models, we should not save the gradients/prev_gradients and activations right? We will have to evaluate them on examples in the forward(-backwards) pass when we load a model to test or train.